tests/nn/data_parallel/test_fsdp_regnet.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+""" Test FSDP with regnet-like model. """
+
+import tempfile
+
+import pytest
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+from torch.nn import BatchNorm2d, Conv2d, Module, SyncBatchNorm
+from torch.optim import SGD
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+from fairscale.nn.data_parallel import TrainingState
+from fairscale.utils.testing import dist_init, skip_if_single_gpu, teardown, torch_version
+
+
+def _test_func(rank, world_size, fsdp_config, tempfile_name, unused):
+    result = dist_init(rank, world_size, tempfile_name, unused)
+    assert result, "Dist init failed"
+
+    assert isinstance(fsdp_config, dict), str(fsdp_config)
+
+    class Model(Module):
+        def __init__(self):
+            super().__init__()
+            # TODO (Min): for now, we just test pytorch sync_bn here.
+            self.conv = Conv2d(2, 2, (1, 1))
+            # Put BN in is own FP32, unflatten, single GPU group.
+            # Note, SyncBN can still have a group size > 1.
+            self.bn = FSDP(
+                BatchNorm2d(2),
+                mixed_precision=False,
+                process_group=dist.new_group(ranks=[rank]),
+                flatten_parameters=False,
+            )
+
+        def forward(self, x):
+            x = self.conv(x)
+            x = self.bn(x)
```

Comment:
is the output type here always float32? Perhaps add:
```python
assert x.dtype == torch.float32
```

Reply:
It is a very good suggestion to add that. The result is something I wasn't expecting. The input to this "FP32 BN" can be either fp16 or fp32 and the output always match the input ind type. The internal weights of BN is kept at FP32. Let me check what kind of precision is in fact required in vision.

Reply:
I also verified that autocast doesn't change this behavior for BN. It seems that we need to explicitly manage it for BN.

Reply:
OK, I found out the answer. I think we only keep the weights fp32 and inputs are still fp16, base on the description here:

<img width="748" alt="image" src="https://user-images.githubusercontent.com/24926999/111559447-c27aee80-874d-11eb-913a-2446e960a530.png">

from this page: https://nvidia.github.io/apex/amp.html